crates/csm-worker/src/processor.rs

```diff
@@ -0,0 +1,163 @@
+    state.last_processed_epoch = Some(epoch);
+
+    Ok(())
+}
+
+/// Update client state with a new checkpoint.
+fn update_client_state_with_checkpoint(
+    state: &mut CsmWorkerState,
+    new_checkpoint: L1Checkpoint,
+    epoch: u64,
+) -> anyhow::Result<()> {
+    // Get the current client state
+    let cur_state = state.cur_state.as_ref();
+
+    // Determine if this checkpoint should be the last finalized or just recent
+    let (last_finalized, recent) = match cur_state.get_last_checkpoint() {
+        Some(existing) => {
+            // If the new checkpoint is for a later epoch, it becomes recent
+            if epoch > existing.batch_info.epoch() {
+                (Some(existing.clone()), Some(new_checkpoint))
+            } else {
+                // Otherwise keep existing
+                (Some(existing.clone()), None)
+            }
+        }
+        None => {
+            // First checkpoint becomes recent
+            (None, Some(new_checkpoint))
+        }
+    };
+
+    // Create new client state
+    let next_state = ClientState::new(last_finalized, recent.clone());
+
+    // Check if we need to finalize an epoch
+    let old_final_epoch = cur_state.get_declared_final_epoch();
+    let new_final_epoch = next_state.get_declared_final_epoch();
+
+    let should_finalize = match (old_final_epoch, new_final_epoch) {
+        (None, Some(new)) => Some(new),
+        (Some(old), Some(new)) if new.epoch() > old.epoch() => Some(new),
+        _ => None,
+    };
+
+    // Store the new client state
+    let l1_block = state.last_asm_block.expect("should have ASM block");
+    state.storage.client_state().put_update_blocking(
+        &l1_block,
+        ClientUpdateOutput::new(next_state.clone(), vec![]),
+    )?;
+
+    // Update our tracked state
+    state.cur_state = Arc::new(next_state);
+
+    // Update status channel
+    state
+        .status_channel
+        .update_client_state(state.cur_state.as_ref().clone(), l1_block);
+
+    // Handle epoch finalization if needed
+    if let Some(epoch_comm) = should_finalize {
+        info!(?epoch_comm, "Finalizing epoch from checkpoint");
+        apply_action(SyncAction::FinalizeEpoch(epoch_comm), &state.storage)?;
+    }
+
+    Ok(())
+}
+
+/// Create a [`Checkpoint`] from a [`CheckpointUpdate`] log.
+///
+/// Note: The log doesn't contain the full signed checkpoint, so we reconstruct
+/// what we can. The signature verification was already done by ASM.
+fn create_checkpoint_from_update(update: &CheckpointUpdate) -> Checkpoint {
+    let epoch = update.batch_info.epoch();
+
+    // Create empty sidecar - checkpoint was already verified by ASM
+    let sidecar = CheckpointSidecar::new(vec![]);
+
+    Checkpoint::new(
+        update.batch_info.clone(),
+        BatchTransition {
+            epoch,
+            chainstate_transition: update.chainstate_transition,
+        },
+        Default::default(), // Empty proof - actual proof was already verified by ASM
```

Comment:
This makes me wonder: why do we need to store historical checkpoints at all?

Reply:
Related: https://github.com/alpenlabs/alpen/pull/1078#discussion_r2399551133